imageio/plugins/pillow.py

```diff
@@ -288,31 +273,54 @@ def write(self, image, *, mode="RGB", format=None, **kwargs):
 
         """
 
-        # ensure that the image has (at least) one batch dimension
-        if image.ndim == 3 and _is_multichannel(mode):
-            image = image[None, ...]
-            save_all = False
-        elif image.ndim == 2 and not _is_multichannel(mode):
-            image = image[None, ...]
-            save_all = False
+        if mode is None:
+            # there is no explicit function in pillow to determine the mode.
+            # However, pillow does it in Image.fromarray, and we imitate that
+            # here
+
+            # pillow supports 1 - 4 color channels depending on the format
+            # and all are channel-last
+            if image.shape[-1] in [2, 3, 4]:
+                shape_key = (1, 1) + image.shape[-1:]
+            else:
+                shape_key = (1, 1)
+
+            try:
+                typekey = shape_key, image.dtype.str
+            except KeyError as e:
+                raise TypeError("Cannot handle this data type") from e
+            try:
+                lookup_mode, _ = Image._fromarray_typemap[typekey]
```

Comment:
This is a bit scary, because it uses a private function, which means that Pillow can change its name or behaviour at any time. I agree that the benefits are probably worth the risk, but a comment would be good.
```suggestion
                # Note: uses private function, so in theory this may break someday.
                lookup_mode, _ = Image._fromarray_typemap[typekey]
```

Reply:
Fair point regarding the use of a private function. Re-reading the code I think I can do a refactor to avoid using it.

Reply:
LGTM!